Source/Core/Core/HW/SystemTimers.cpp

```diff
@@ -206,22 +177,101 @@ static void ThrottleCallback(u64 last_time, s64 cyclesLate)
   const SConfig& config = SConfig::GetInstance();
   bool frame_limiter = config.m_EmulationSpeed > 0.0f && !Core::GetIsThrottlerTempDisabled();
   u32 next_event = GetTicksPerSecond() / 1000;
+
+  {
+    std::lock_guard<std::mutex> lk(s_emu_to_real_time_mutex);
+    s_emu_to_real_time_ring_buffer[s_emu_to_real_time_index] = time - s_time_spent_sleeping;
+    s_emu_to_real_time_index =
+        (s_emu_to_real_time_index + 1) % s_emu_to_real_time_ring_buffer.size();
+  }
+
   if (frame_limiter)
   {
     if (config.m_EmulationSpeed != 1.0f)
       next_event = u32(next_event * config.m_EmulationSpeed);
     const s64 max_fallback = config.iTimingVariance * 1000;
     if (abs(diff) > max_fallback)
     {
-      DEBUG_LOG(COMMON, "system too %s, %d ms skipped", diff < 0 ? "slow" : "fast",
+      DEBUG_LOG(COMMON, "system too %s, %ld ms skipped", diff < 0 ? "slow" : "fast",
                 abs(diff) - max_fallback);
       last_time = time - max_fallback;
     }
-    else if ((diff / 1000) > 0)
+    else if (diff > 1000)
+    {
       Common::SleepCurrentThread(diff / 1000);
+      s_time_spent_sleeping += Common::Timer::GetTimeUs() - time;
+    }
   }
   CoreTiming::ScheduleEvent(next_event - cyclesLate, et_Throttle, last_time + 1000);
 }
+}  // namespace
+
+u32 GetTicksPerSecond()
+{
+  return s_cpu_core_clock;
+}
+
+void DecrementerSet()
+{
+  u32 decValue = PowerPC::ppcState.spr[SPR_DEC];
+
+  CoreTiming::RemoveEvent(et_Dec);
+  if ((decValue & 0x80000000) == 0)
+  {
+    CoreTiming::SetFakeDecStartTicks(CoreTiming::GetTicks());
+    CoreTiming::SetFakeDecStartValue(decValue);
+
+    CoreTiming::ScheduleEvent(decValue * TIMER_RATIO, et_Dec);
+  }
+}
+
+u32 GetFakeDecrementer()
+{
+  return (CoreTiming::GetFakeDecStartValue() -
+          (u32)((CoreTiming::GetTicks() - CoreTiming::GetFakeDecStartTicks()) / TIMER_RATIO));
+}
+
+void TimeBaseSet()
+{
+  CoreTiming::SetFakeTBStartTicks(CoreTiming::GetTicks());
+  CoreTiming::SetFakeTBStartValue(PowerPC::ReadFullTimeBaseValue());
+}
+
+u64 GetFakeTimeBase()
+{
+  return CoreTiming::GetFakeTBStartValue() +
+         ((CoreTiming::GetTicks() - CoreTiming::GetFakeTBStartTicks()) / TIMER_RATIO);
+}
+
+s64 GetLocalTimeRTCOffset()
+{
+  return s_localtime_rtc_offset;
+}
+
+double GetEstimatedEmulationPerformance()
+{
+  u64 ts_now, ts_before;  // In microseconds
+  {
+    std::lock_guard<std::mutex> lk(s_emu_to_real_time_mutex);
+    size_t index_now = s_emu_to_real_time_index == 0 ? s_emu_to_real_time_ring_buffer.size() :
+                                                       s_emu_to_real_time_index - 1;
+    size_t index_before = s_emu_to_real_time_index;
```

Comment:
This seems odd to me. Values are written to `s_emu_to_real_time_index`, then it is incremented by 1 (wrapping around if necessary). So `index_now` should be `s_emu_to_real_time_index - 1` (+/-wrapping), and `index_before` should be...`index_now - 1`? Or is the name `index_before` just there to confuse me? Before _what_ exactly?

Reply:
Yeah, I'm open to better naming. index_before is the index of the first value in the ring buffer, 1000ms ago. I call it before because of the (now - before) computation. Since it's the first value, it's the one we'd write to next, hence index_now + 1 instead of -1.